File: xuri-rpc-test/testscript/pt1-client.js
import { PlainProxyManager,RunnableProxyManager,MessageReceiver,Client,asProxy,getMessageReceiver,setHostId } from 'xuri-rpc'
import {WebSocket} from 'ws'
import { assertV } from './base/util.js'


//define a sender
class Sender {
  constructor(ws) {
    this.ws=ws
  }
  async send(message) {
    this.ws.send(JSON.stringify(message))
  }
}
async function main(){

    setHostId('frontend')
    let client=new Client()
    
    
    const ws = new WebSocket('ws://localhost:18081');
    await new Promise((resolve, reject) => {
      ws.on('open', resolve);
      ws.on('error', reject);
    });
    ws.on('message', (data) => {
      getMessageReceiver().onReceiveMessage(JSON.parse(data),client)
      console.log(`收到服务器消息: ${data}`);
    });
    client.setSender(new Sender(ws))
    

    let main=await client.getMain()
    let result=await main.plus(1,2,asProxy((result)=>console.log('from callback',result)))
    console.log('from rpc',result)
    assertV(result==3,'error')
    process.exit(0)
}
main()

File: xuri-rpc-test/testscript/pt1-client.py
import asyncio
import json
import websockets
from xuri_rpc import PlainProxyManager, RunnableProxyManager, MessageReceiver, Client, asProxy, getMessageReceiver, setHostId
setHostId('backend')
from xuri_rpc import setDebugFlag
setDebugFlag(True)


# define a sender
class Sender:
    def __init__(self, ws):
        self.ws = ws

    async def send(self, message):
        # message is an object can be jsonified
        await self.ws.send(json.dumps(message))

async def main():
    setHostId('frontend')
    client = Client()

    ws = await websockets.connect('ws://localhost:18081')

    async def on_message(data):
        await getMessageReceiver().onReceiveMessage(json.loads(data), client)
        print(f'收到服务器消息: {data}')

    # Run message reception in the background
    async def listen():
        async for message in ws:
            asyncio.ensure_future(on_message(message))

    # Start listening in the background
    asyncio.ensure_future(listen())

    client.setSender(Sender(ws))

    main_proxy = await client.getMain()
    def callback(result):
        # breakpoint()
        print('from callback', result)
    result = await main_proxy.plus(1, 2, asProxy(callback))
    print('from rpc', result)

asyncio.run(main())